rust-guest/src/context/context.rs

```diff
@@ -0,0 +1,102 @@
+#![allow(dead_code)]
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+
+use std::io::Cursor;
+use std::mem;
+use std::collections::HashMap;
+use super::generated::{Encode, Decode, Context, Request, Response};
+
+lazy_static! {
+    pub static ref PTR_LEN: Mutex<(u32, u32)> = Mutex::new((0, 0));
+}
+
+pub trait RunContext {
+    fn new() -> Self;
+    fn from_read_buffer(self, read_buff_global: &mut Cursor<&mut Vec<u8>>) -> Self;
+    fn to_write_buffer(self) -> (u32, u32);
+    fn next(self) -> Self;
+    fn request(&mut self) -> &mut Request;
+    fn response(&mut self) -> &mut Response;
+    fn default() -> Self;
+}
+
+impl RunContext for Context {
+    fn new()  -> Context {
+          Self::default()
+    }
+
+    fn from_read_buffer(self, read_buff_global: &mut Cursor<&mut Vec<u8>>) -> Self {
+          Decode::decode(read_buff_global).unwrap().unwrap()
+    }
+
+    fn to_write_buffer(self) -> (u32, u32) {
+        let mut cursor = Cursor::new(Vec::new());
+        let _ = Encode::encode(self, &mut cursor);
+
+        let mut vec = cursor.into_inner();
+        vec.shrink_to_fit();
+        let ptr = vec.as_ptr() as u32;
+        let len = vec.len() as u32;
+        mem::forget(vec);  // prevents deallocation in Rust
+                           // vec still exists in mem, but
+                           // rust doesn't have any concept of it
+
+        return (ptr, len)
+    }
+
+    fn request(&mut self) -> &mut Request {
+        &mut self.request
+    }
+
+    fn response(&mut self) -> &mut Response {
+        &mut self.response
+    }
+
+    fn next(self) -> Self {
+         unsafe {
```

Comment:
How much of this function needs to be unsafe?

Reply:
Not the entire thing, but unfortunately anything defined within the unsafe block is out of scope for the remainder of the func, so can call write_to_buffer() and then the rest of the function needs to be wrapped in the unsafe.